File: firmware/pathfinder.py
"""
pathfinder.py — Lightweight Dijkstra for ESP32 (MicroPython)
Optimised for minimal heap usage on 520KB SRAM.
"""

from config import NODES, EDGES, EXIT_NODES


def _build_adjacency(nodes, edges):
    """Build adjacency list: {node_id: [(neighbour_id, base_weight)]}"""
    adj = {nid: [] for nid in nodes}
    for (frm, to, w) in edges:
        adj[frm].append((to, w))
        adj[to].append((frm, w))
    return adj


# Pre-build adjacency at module load (runs once on boot)
_ADJ = _build_adjacency(NODES, EDGES)


class _MinHeap:
    """Minimal binary min-heap for MicroPython (no heapq in all builds)."""
    def __init__(self):
        self._h = []

    def push(self, priority, value):
        self._h.append((priority, value))
        self._bubble_up(len(self._h) - 1)

    def pop(self):
        if not self._h:
            return None
        top = self._h[0]
        last = self._h.pop()
        if self._h:
            self._h[0] = last
            self._sink_down(0)
        return top

    def __len__(self):
        return len(self._h)

    def _bubble_up(self, i):
        while i > 0:
            p = (i - 1) >> 1
            if self._h[p][0] <= self._h[i][0]:
                break
            self._h[p], self._h[i] = self._h[i], self._h[p]
            i = p

    def _sink_down(self, i):
        n = len(self._h)
        while True:
            sm = i
            l, r = 2*i+1, 2*i+2
            if l < n and self._h[l][0] < self._h[sm][0]: sm = l
            if r < n and self._h[r][0] < self._h[sm][0]: sm = r
            if sm == i: break
            self._h[i], self._h[sm] = self._h[sm], self._h[i]
            i = sm


def dijkstra(source_id, weights=None):
    """
    Run Dijkstra from source_id.

    Args:
        source_id: str — starting node
        weights:   dict {(from,to): float} — live fused weights.
                   If None, uses base weights from EDGES.
                   Infinity means edge is blocked.

    Returns:
        (dist: dict, prev: dict)
    """
    dist    = {nid: float('inf') for nid in NODES}
    prev    = {nid: None for nid in NODES}
    visited = set()
    pq      = _MinHeap()

    dist[source_id] = 0.0
    pq.push(0.0, source_id)

    while len(pq) > 0:
        cost, u = pq.pop()
        if u in visited:
            continue
        visited.add(u)

        for (v, base_w) in _ADJ.get(u, []):
            # Prefer live fused weights if provided
            if weights is not None:
                w = weights.get((u, v), weights.get((v, u), base_w))
            else:
                w = base_w

            if w == float('inf'):
                continue

            nd = cost + w
            if nd < dist[v]:
                dist[v]  = nd
                prev[v]  = u
                pq.push(nd, v)

    return dist, prev


def _reconstruct(end_id, prev):
    """Reconstruct the node path list from Dijkstra prev map."""
    path = []
    cursor = end_id
    while cursor is not None:
        path.append(cursor)
        cursor = prev[cursor]
    path.reverse()
    return path


def find_evacuation_path(source_id, weights=None):
    """
    Find safest path from source_id to the nearest exit.

    Returns:
        {'path': [node_id, ...], 'cost': float, 'exit': str} | None
    """
    if source_id not in NODES:
        print(f'[Pathfinder] Unknown node: {source_id}')
        return None

    import utime
    t0 = utime.ticks_ms()

    dist, prev = dijkstra(source_id, weights)

    best_exit = None
    best_cost = float('inf')
    for eid in EXIT_NODES:
        c = dist.get(eid, float('inf'))
        if c < best_cost:
            best_cost = c
            best_exit = eid

    elapsed = utime.ticks_diff(utime.ticks_ms(), t0)
    if elapsed > 300:
        print(f'[Pathfinder] WARNING: took {elapsed}ms (must be <300ms)')

    if best_exit is None or best_cost == float('inf'):
        print('[Pathfinder] No reachable exit found!')
        return None

    path = _reconstruct(best_exit, prev)
    return {'path': path, 'cost': best_cost, 'exit': best_exit}


def find_fallback_path(source_id):
    """Use base weights only (no sensor data) — for fail-safe mode."""
    return find_evacuation_path(source_id, weights=None)

File: dashboard/src/core/pathfinder.ts
// ============================================================
// Dijkstra Pathfinding Engine
// Finds shortest safe evacuation path from any node to exits
// ============================================================

import type { BuildingGraph, EvacuationPath, HazardLevel } from './types';
import { EXIT_NODE_IDS } from './building-layout';

// ─── Min-Heap (Binary Heap) Priority Queue ──────────────────
class MinHeap<T> {
  private heap: { priority: number; value: T }[] = [];

  push(priority: number, value: T) {
    this.heap.push({ priority, value });
    this._bubbleUp(this.heap.length - 1);
  }

  pop(): { priority: number; value: T } | undefined {
    if (this.heap.length === 0) return undefined;
    const top = this.heap[0];
    const last = this.heap.pop()!;
    if (this.heap.length > 0) {
      this.heap[0] = last;
      this._sinkDown(0);
    }
    return top;
  }

  get size() { return this.heap.length; }

  private _bubbleUp(i: number) {
    while (i > 0) {
      const parent = Math.floor((i - 1) / 2);
      if (this.heap[parent].priority <= this.heap[i].priority) break;
      [this.heap[parent], this.heap[i]] = [this.heap[i], this.heap[parent]];
      i = parent;
    }
  }

  private _sinkDown(i: number) {
    const n = this.heap.length;
    while (true) {
      let smallest = i;
      const l = 2 * i + 1, r = 2 * i + 2;
      if (l < n && this.heap[l].priority < this.heap[smallest].priority) smallest = l;
      if (r < n && this.heap[r].priority < this.heap[smallest].priority) smallest = r;
      if (smallest === i) break;
      [this.heap[smallest], this.heap[i]] = [this.heap[i], this.heap[smallest]];
      i = smallest;
    }
  }
}

// ─── Dijkstra Core ──────────────────────────────────────────
interface DijkstraResult {
  dist: Map<string, number>;
  prev: Map<string, string | null>;   // nodeId -> came-from nodeId
  prevEdge: Map<string, string | null>; // nodeId -> came-from edgeId
}

function dijkstra(graph: BuildingGraph, sourceId: string): DijkstraResult {
  const dist    = new Map<string, number>();
  const prev    = new Map<string, string | null>();
  const prevEdge= new Map<string, string | null>();
  const visited = new Set<string>();
  const pq      = new MinHeap<string>();

  for (const nodeId of graph.nodes.keys()) {
    dist.set(nodeId, Infinity);
    prev.set(nodeId, null);
    prevEdge.set(nodeId, null);
  }

  dist.set(sourceId, 0);
  pq.push(0, sourceId);

  while (pq.size > 0) {
    const { priority: cost, value: u } = pq.pop()!;
    if (visited.has(u)) continue;
    visited.add(u);

    const edgeIds = graph.adjacency.get(u) ?? [];
    for (const edgeId of edgeIds) {
      const edge = graph.edges.get(edgeId)!;
      if (edge.isBlocked || edge.currentWeight === Infinity) continue;

      const v = edge.from === u ? edge.to : edge.from;
      const newDist = cost + edge.currentWeight;

      if (newDist < (dist.get(v) ?? Infinity)) {
        dist.set(v, newDist);
        prev.set(v, u);
        prevEdge.set(v, edgeId);
        pq.push(newDist, v);
      }
    }
  }

  return { dist, prev, prevEdge };
}

// ─── Reconstruct path from Dijkstra result ──────────────────
function reconstructPath(
  endId: string,
  prev: Map<string, string | null>,
  prevEdge: Map<string, string | null>,
  dist: Map<string, number>,
  graph: BuildingGraph,
): EvacuationPath | null {
  if ((dist.get(endId) ?? Infinity) === Infinity) return null;

  const nodeIds: string[] = [];
  const edgeIds: string[] = [];
  let cursor = endId;

  while (cursor !== null && cursor !== undefined) {
    nodeIds.unshift(cursor);
    const edge = prevEdge.get(cursor);
    if (edge) edgeIds.unshift(edge);
    cursor = prev.get(cursor)!;
  }

  // Determine worst hazard along path
  let worstHazard: HazardLevel = 'SAFE';
  const levels: HazardLevel[] = ['SAFE', 'CAUTION', 'WARNING', 'DANGER', 'BLOCKED'];
  for (const eid of edgeIds) {
    const edge = graph.edges.get(eid);
    if (!edge) continue;
    const idx = levels.indexOf(edge.hazardLevel);
    if (idx > levels.indexOf(worstHazard)) worstHazard = edge.hazardLevel;
  }

  return {
    nodeIds,
    edgeIds,
    totalCost: dist.get(endId) ?? Infinity,
    exitNodeId: endId,
    hazardLevel: worstHazard,
    computedAt: Date.now(),
  };
}

// ─── Public API ─────────────────────────────────────────────

/**
 * Find the safest evacuation path from `sourceId` to any exit.
 * Returns the path with lowest total fused cost (safest route).
 */
export function findEvacuationPath(
  graph: BuildingGraph,
  sourceId: string,
  exitIds: string[] = EXIT_NODE_IDS,
): EvacuationPath | null {
  if (!graph.nodes.has(sourceId)) {
    console.error(`[Pathfinder] Unknown source node: ${sourceId}`);
    return null;
  }

  const start = performance.now();
  const { dist, prev, prevEdge } = dijkstra(graph, sourceId);

  // Pick the exit with lowest cost
  let bestPath: EvacuationPath | null = null;
  for (const exitId of exitIds) {
    const path = reconstructPath(exitId, prev, prevEdge, dist, graph);
    if (!path) continue;
    if (!bestPath || path.totalCost < bestPath.totalCost) {
      bestPath = path;
    }
  }

  const elapsed = performance.now() - start;
  if (elapsed > 300) {
    console.warn(`[Pathfinder] Computation took ${elapsed.toFixed(1)}ms (must be < 300ms)`);
  }

  return bestPath;
}

/**
 * Compute paths from every non-exit node to the safest exit.
 * Returns a map: sourceNodeId → EvacuationPath.
 */
export function computeAllPaths(
  graph: BuildingGraph,
  exitIds: string[] = EXIT_NODE_IDS,
): Map<string, EvacuationPath> {
  const result = new Map<string, EvacuationPath>();

  for (const [nodeId, node] of graph.nodes) {
    if (node.isExit) continue;
    const path = findEvacuationPath(graph, nodeId, exitIds);
    if (path) result.set(nodeId, path);
  }

  return result;
}

/**
 * Fallback: compute path purely on base weights (ignoring sensor data).
 * Used when communication is lost and we need a safe default.
 */
export function findFallbackPath(
  graph: BuildingGraph,
  sourceId: string,
  exitIds: string[] = EXIT_NODE_IDS,
): EvacuationPath | null {
  // Temporarily override weights with base weights
  const tempGraph: BuildingGraph = {
    nodes: graph.nodes,
    adjacency: graph.adjacency,
    edges: new Map(
      Array.from(graph.edges.entries()).map(([id, edge]) => [
        id,
        { ...edge, currentWeight: edge.baseWeight, isBlocked: false },
      ]),
    ),
  };
  return findEvacuationPath(tempGraph, sourceId, exitIds);
}
